### PR TITLE
Clean up CMake

### PR DESCRIPTION
**Pull request by matthew-d-jones**
### Issue

DM-1555

### Description of work

- Clean up CMake
- Update dependencies - created ticket DM-1565 to update librdkafka as requires more work.
- Make unit tests build on Windows - note, they can be run on a developer machine on Windows, but they cause popup error dialogues to report that CARepeater is not running. As this would probably cause a problem on the Jenkins node I have not added running the tests to the Windows pipeline yet.

graylog-logger being missing is now the only difference in the Windows build.

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -2,63 +2,56 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ..)
...
+        Threads::Threads
 )
 
-if (UNIX)
-  list(APPEND libraries_common pthread)
-endif()
+set(path_include_common "")
+
+# Allow a system installed EPICS library to be used instead of conan package (for ISIS-STFC)
+set(USE_EXTERNAL_EPICS FALSE CACHE BOOL "Look for system installed EPICS instead of using Conan package")
```
Comment (SkyToGround) [inline src/CMakeLists.txt]: I think we talked about this a while ago. Is there no possibility of getting STFC (Freddie I assume) to start using conan?

File: src/Forwarder.cpp [hunk 2]
```diff
@@ -39,6 +39,14 @@ std::vector<char> getHostname() {
 #endif
 
 #include "CURLReporter.h"
+#include "schemas/f142/f142.cpp"
+
+namespace {
+void registerSchemas() {
```
Comment (SkyToGround) [inline src/Forwarder.cpp]: Considering that we only have one module in the forwarder (I should REALLY finish that pull request for the chopper module), I do not mind putting this as C++ code. However, I am still interested in hearing if you have any reasoning to do this as it has been working ore or less as intended.

File: src/CMakeLists.txt (hunk 1 above)
Comment (matthew-d-jones) [reply src/CMakeLists.txt]: Yes, we're using Conan for everything other than EPICS. The controls team want to use exactly the same version of EPICS for everything. Also several applications run on one machine on each instrument, and EPICS is bigger than all the other libraries combined so it does make the package to deploy a lot smaller without it.

File: src/tests/CMakeLists.txt [hunk 3]
```diff
@@ -14,16 +13,19 @@ set(sources
...
     ProducerDeliveryCbTests.cpp
     ConsumerTests.cpp
     MockMessage.h
     UnitTests.cpp)
-add_executable(${tgt} ${sources})
-add_dependencies(${tgt} flatbuffers_generate)
-target_include_directories(${tgt} PRIVATE ${path_include_common})
-target_link_libraries(${tgt} ${libraries_common})
-add_gtest_to_target(${tgt})
+add_executable(tests ${sources})
```
Comment (SkyToGround) [inline src/tests/CMakeLists.txt]: I really like that you have removed this `${tgt}`-nonsense. With that said, I do prefer to use a (de-facto standard) `*_SRC` name.
```suggestion
add_executable(tests ${tests_SRC})
```

File: src/tests/CMakeLists.txt [hunk 4]
```diff
@@ -1,6 +1,5 @@
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ../../tests)
 
-set(tgt "tests")
 set(sources
```
Comment (SkyToGround) [inline src/tests/CMakeLists.txt]:
```suggestion
set(tests_SRC
```

File: src/Forwarder.cpp (hunk 2 above)
Comment (matthew-d-jones) [reply src/Forwarder.cpp]: After my CMake changes the module was failing to be registered and it was not obvious to me how the magic it was using before was working, I thought that was reason enough to make it explicit.

Comment (SkyToGround) [reply src/Forwarder.cpp]: Ok, that is a good enough reason for me.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -139,49 +132,30 @@ set(SOURCES
     Timer.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/git_commit_current.cpp)
 
-set(tgt __objects)
-add_library(${tgt} OBJECT ${SOURCES} ${INCLUDES})
-target_include_directories(${tgt} PRIVATE ${path_include_common})
-target_include_directories(${tgt} SYSTEM PRIVATE ${path_include_common_suppressed_warnings})
-
+add_library(forwarder-library ${SOURCES} ${INCLUDES})
```
Comment (SkyToGround) [inline src/CMakeLists.txt]: I can see why the automatic registration of modules stopped working. You no longer use an `OBJECT` library. That is fine with me though.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -2,63 +2,56 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ..)
...
-    message(WARNING "graylog_logger requested but not found")
-endif()
+    list(APPEND libraries_common
+            CONAN_PKG::spdlog-graylog)
+else()
+    list(APPEND libraries_common
+            CONAN_PKG::spdlog)
+endif(CONAN_SPDLOG-GRAYLOG_ROOT AND USE_GRAYLOG_LOGGER)
 
 set(INCLUDES
```
Comment (SkyToGround) [inline src/CMakeLists.txt]: I very much prefer a (somewhat) clear indication of to which target a set of source files belong.
```suggestion
set(forwarder-library_INC
```

File: src/CMakeLists.txt (hunk 5 above)
Comment (SkyToGround) [inline src/CMakeLists.txt]: See previous comment.
```suggestion
add_library(forwarder-library ${forwarder-library_SRC} ${forwarder-library_INC})
```
Also requires fixing of the `set(SOURCES`-line.

**Comment by SkyToGround** [comment]
@matthew-d-jones I have done a first quick review but I want to take a closer look at the changes. I  will try to do that this afternoon.

File: Dockerfile [hunk 7]
```diff
@@ -30,7 +30,7 @@ RUN export DEBIAN_FRONTEND=noninteractive \
 COPY ./ ../forwarder_src/
 
 RUN cd forwarder \
-    && cmake -DCONAN="MANUAL" --target="forward-epics-to-kafka" -DGOOGLETEST_DISABLE="ON" ../forwarder_src \
+    && cmake -DCONAN="MANUAL" --target="forward-epics-to-kafka" -DBUILD_TESTS="OFF" ../forwarder_src \
```
Comment (SkyToGround) [inline Dockerfile]: 👍

**Pull request merged**